feat: add resource interpreter customization for SidecarSet

abhi0324 opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds resource interpreter customization for OpenKruise SidecarSet, enabling Karmada to properly manage SidecarSet resources in multi-cluster environments. This includes:

- Status aggregation across member clusters
- Health checks for SidecarSet resources
- Dependency interpretation for ConfigMaps and Secrets
- Comprehensive test coverage

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:
- Includes test cases for multi-cluster scenarios
- Verified with `karmadactl interpret` command
- Follows the same pattern as other resource interpreter customizations

**Does this PR introduce a user-facing change?**:
```release-note
feat: Add resource interpreter support for OpenKruise SidecarSet

Adds support for interpreting OpenKruise SidecarSet resources in Karmada, including:
- Multi-cluster status aggregation
- Health checks
- Dependency resolution for ConfigMaps and Secrets
- Comprehensive test coverage

Users can now properly manage SidecarSet resources across multiple clusters with Karmada.

RainbowMango commented:
Thank you, @abhi0324, for doing this. 
Have you tested it? Or could you please share how to test it? It might be needed for the reviewer to run the test locally.

XiShanYongYe-Chang commented on pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/SidecarSet/customizations.yaml:
```diff
@@ -0,0 +1,149 @@
+              desiredObj.spec.volumes = observedObj.spec.volumes
+            end
+          end
+          -- Always preserve status
+          if observedObj.status ~= nil then
+            desiredObj.status = observedObj.status
+          end
+          return desiredObj
```
The status will not be overwritten during the retain operation, so there should be no need to declare this logic again here.

abhi0324 replied:
@XiShanYongYe-Chang Thank you for the reminder! I've now removed the redundant status preservation logic in the latest commit.

XiShanYongYe-Chang commented on pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/SidecarSet/customizations.yaml:
```diff
@@ -0,0 +1,149 @@
+            end
+          end
+          return dependencies
+        end 
```
Can you help add a new blank at the end of the file. Same as other files.

XiShanYongYe-Chang reviewed (commented):
Thanks a lot~

abhi0324 commented:
Hey @yike21 , here are the test result:

```
---
# [1/1] aggregatedStatus:
apiVersion: apps.kruise.io/v1alpha1
kind: SidecarSet
metadata:
    generation: 1
    labels:
        app: sample
    name: sample-sidecarset
    namespace: test-sidecarset
status:
    matchedPods: 3
    readyPods: 3
    updatedPods: 3
---
# [1/1] dependencies:
- apiVersion: v1
  kind: ConfigMap
  name: sidecar-config
  namespace: test-sidecarset
- apiVersion: v1
  kind: Secret
  name: sidecar-secret
  namespace: test-sidecarset
---
# [1/1] healthy:
true
---
# [1/2] replica:
0
---
# [2/2] requires:
null
---
# [1/1] retained:
apiVersion: apps.kruise.io/v1alpha1
kind: SidecarSet
metadata:
    generation: 1
    labels:
        app: sample
    name: sample-sidecarset
    namespace: test-sidecarset
spec:
    containers:
        - env:
            - name: CONFIG_DATA
              valueFrom:
                configMapKeyRef:
                    key: config
                    name: sidecar-config
            - name: SECRET_DATA
              valueFrom:
                secretKeyRef:
                    key: secret
                    name: sidecar-secret
          image: busybox:latest
          name: sidecar
    injectionStrategy: BeforeAppContainer
    selector:
        matchLabels:
            app: sample
            test: sidecarset
    volumes:
        - configMap:
            name: sidecar-config
          name: configmap
        - name: secret
          secret:
            secretName: sidecar-secret
-----------------------------------
SOURCE: declarative-configuration-sidecarset
TARGET: apps.kruise.io/v1alpha1 SidecarSet
RULERS:
    Retain:                                  PASS
    InterpretReplica:                        PASS
    ReviseReplica:                           UNSET
    InterpretStatus:                         UNSET
    AggregateStatus:                         PASS
    InterpretHealth:                         PASS
    InterpretDependency:                     PASS

abhi0324 commented:
Hey @XiShanYongYe-Chang , I have done the required changes

yike21 commented:
Thanks a lot~ @abhi0324 Could you squash the commits under this PR into one? and then rebase the latest master commit.

XiShanYongYe-Chang commented on pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/SidecarSet/customizations.yaml:
```diff
@@ -0,0 +1,145 @@
+            }
+          }
+        end
+    retention:
+      luaScript: |
+        function Retain(desiredObj, observedObj)
+          -- Preserve injectionStrategy if it exists
+          if observedObj.spec and observedObj.spec.injectionStrategy then
+            if not desiredObj.spec then
+              desiredObj.spec = {}
+            end
+            desiredObj.spec.injectionStrategy = observedObj.spec.injectionStrategy
+          end
+          -- Preserve other runtime fields that shouldn't be overwritten
+          if observedObj.spec and observedObj.spec.volumes then
+            if not desiredObj.spec then
+              desiredObj.spec = {}
+            end
+            -- Only preserve if not explicitly set in desired
+            if not desiredObj.spec.volumes then
+              desiredObj.spec.volumes = observedObj.spec.volumes
+            end
+          end
+          return desiredObj
+        end
```
Once the resources are propagated to the member clusters, are these two fields then controlled by the controllers within the member clusters? It might be necessary to explain why these two fields are retained to ensure that this logic is sufficiently general.

XiShanYongYe-Chang replied:
Thanks a lot~ @mohamedawnallah 

This question may require the author to provide an answer.

Hi @abhi0324, can you help answer it?

XiShanYongYe-Chang commented on pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/SidecarSet/customizations.yaml:
```diff
@@ -0,0 +1,145 @@
+                      name = env.valueFrom.configMapKeyRef.name,
+                      namespace = desiredObj.metadata.namespace
+                    })
+                  end
+                  if env.valueFrom and env.valueFrom.secretKeyRef then
+                    table.insert(dependencies, {
+                      apiVersion = "v1",
+                      kind = "Secret",
+                      name = env.valueFrom.secretKeyRef.name,
+                      namespace = desiredObj.metadata.namespace
+                    })
+                  end
+                end
+              end
+            end
+          end
+          -- Check for volume references
+          if desiredObj.spec and desiredObj.spec.volumes then
+            for i = 1, #desiredObj.spec.volumes do
+              local volume = desiredObj.spec.volumes[i]
+              if volume.configMap then
+                table.insert(dependencies, {
+                  apiVersion = "v1",
+                  kind = "ConfigMap",
+                  name = volume.configMap.name,
+                  namespace = desiredObj.metadata.namespace
+                })
+              end
+              if volume.secret then
+                table.insert(dependencies, {
+                  apiVersion = "v1",
+                  kind = "Secret",
+                  name = volume.secret.secretName,
+                  namespace = desiredObj.metadata.namespace
+                })
+              end
+            end
+          end
+          return dependencies
+        end 
```
It appears that the checks for `envFrom` in the container and `Volume.project` are meaningful.

XiShanYongYe-Chang replied:
Similar Go language code:

https://github.com/karmada-io/karmada/blob/59c2837c1bcb7cd7f63f03a4191192d2b83872e0/pkg/util/lifted/visitpod.go#L233-L239

https://github.com/karmada-io/karmada/blob/59c2837c1bcb7cd7f63f03a4191192d2b83872e0/pkg/util/lifted/visitpod.go#L212-L217

abhi0324 commented:
@XiShanYongYe-Chang  Thank you for your feedback! I've made the following improvements:

### 1. Retention Logic
I have updated the retention logic in `customizations.yaml` for SidecarSet.  
- **Previous behavior:** The function retained `injectionStrategy` and `volumes` from the observed object, which could prevent user-driven updates or deletions via Karmada.
- **Current behavior:** The retention function now simply returns the desired object, allowing Karmada users to fully manage all configuration fields (including `injectionStrategy` and `volumes`). This ensures users can add, update, or remove these fields as expected, and addresses the concern about blocking user intent.

### 2. Dependency Interpretation
I have also updated the dependency interpretation logic to ensure all relevant dependencies are captured:
- The `GetDependencies` function now checks for:
  - `env.valueFrom` (ConfigMap and Secret references)
  - `envFrom` (ConfigMapRef and SecretRef)
  - Standard volume references (`configMap`, `secret`)
  - Projected volumes and their sources (ConfigMap and Secret)
- This matches the logic in Karmada's Go codebase and ensures all dependencies are accounted for, preventing potential application failures due to missing resources.

The changes are now ready for review. Please let me know if you have any other suggestions!

XiShanYongYe-Chang commented:
/gemini review

XiShanYongYe-Chang commented:
Thanks~ @abhi0324 
Ask the module owner to help review again.
/assign @yike21

XiShanYongYe-Chang commented:
/lgtm

XiShanYongYe-Chang commented:
Hi @yike21, @RainbowMango would you like to help take a review?
/assign @yike21 @RainbowMango

RainbowMango commented:
@yike21 Do you have any further comments?

yike21 commented:
/approve
Nice coding style. Thanks a lot! @abhi0324
